pkg/controller/common.go

```diff
@@ -366,51 +389,96 @@ func (c *SharedState) deleteVolumeAndClaim(azVolumeName string) {
 	}
 }
 
-func selectNodes(ctx context.Context, azr azReconciler, numNodes int, primaryNode string, exceptionNodes map[string]int) ([]string, error) {
-	// TODO: fill with node selection logic (based on node capacity and node failure domain)
-	/*
-		Below is temporary node selection logic
-	*/
+func getNodeRemainingCapacity(ctx context.Context, cachedClient client.Client, nodeObj *v1.Node, nodeName string) (int, error) {
+	// get node if necessary
+	if nodeObj == nil {
+		nodeObj = &v1.Node{}
+		if err := cachedClient.Get(ctx, types.NamespacedName{Name: nodeName}, nodeObj); err != nil {
+			return -1, err
+		}
+	}
+
+	// get all replica azvolumeattachments on the node
+	replicaAttachments, err := getAzVolumeAttachmentsForNode(ctx, cachedClient, nodeName, replicaOnly)
+	if err != nil {
+		return -1, err
+	}
+
+	// check node capacity
+	maxAttachables, ok := nodeObj.Status.Allocatable[consts.AttachableVolumesField]
```

Comment:
Are we sure this field is populated properly? This is an optional field. We might need to rely on the vm capacity(per its type) to determine the max number of attachments.